pkg/hubble/metrics/metrics_test.go

```diff
@@ -123,3 +101,259 @@ func TestHubbleServerStandalone(t *testing.T) {
+	}
+}
+
+func TestHandlersUpdatedInDfpOnConfigChange(t *testing.T) {
+	// Handlers: +drop
+	watcher := metricConfigWatcher{configFilePath: "testdata/valid_metric_config.yaml"}
+	cfg, _, _, err := watcher.readConfig()
+	require.Nil(t, err)
+
+	reg := prometheus.NewPedanticRegistry()
+	dfp := DynamicFlowProcessor{registry: reg}
+	assert.Nil(t, dfp.Metrics)
+	dfp.onConfigReload(context.TODO(), false, 0, *cfg)
+	assertHandlersInDfp(t, &dfp, cfg)
+
+	// Handlers: =drop, +flow
+	watcher = metricConfigWatcher{configFilePath: "testdata/valid_metric_config_2.yaml"}
+	cfg, _, _, err = watcher.readConfig()
+	require.Nil(t, err)
+
+	dfp.onConfigReload(context.TODO(), false, 0, *cfg)
+	assertHandlersInDfp(t, &dfp, cfg)
+
+	// Handlers: -drop, =flow
+	watcher = metricConfigWatcher{configFilePath: "testdata/valid_metric_config_3.yaml"}
+	cfg, _, _, err = watcher.readConfig()
+	require.Nil(t, err)
+
+	dfp.onConfigReload(context.TODO(), false, 0, *cfg)
+	assertHandlersInDfp(t, &dfp, cfg)
+}
+
+func assertHandlersInDfp(t *testing.T, dfp *DynamicFlowProcessor, cfg *api.Config) {
+	names := cfg.GetMetricNames()
+	assert.Equal(t, len(names), len(dfp.Metrics.Handlers))
+	for _, m := range dfp.Metrics.Handlers {
+		_, ok := names[m.Name]
+		assert.True(t, ok)
+	}
+}
+
+func TestMetricReRegisterAndCollect(t *testing.T) {
+	// Handlers: +drop
+	watcher := metricConfigWatcher{configFilePath: "testdata/valid_metric_config.yaml"}
+	cfg, _, _, err := watcher.readConfig()
+	require.Nil(t, err)
+
+	reg := prometheus.NewPedanticRegistry()
+	dfp := DynamicFlowProcessor{registry: reg}
+	dfp.onConfigReload(context.TODO(), false, 0, *cfg)
+
+	flow1 := &pb.Flow{
+		EventType: &pb.CiliumEventType{Type: monitorAPI.MessageTypePolicyVerdict},
+		L4: &pb.Layer4{
+			Protocol: &pb.Layer4_TCP{
+				TCP: &pb.TCP{},
+			},
+		},
+		Source:         &pb.Endpoint{Namespace: "allow", PodName: "src_pod"},
+		Destination:    &pb.Endpoint{Namespace: "allow", PodName: "dst_pod"},
+		Verdict:        pb.Verdict_DROPPED,
+		DropReason:     uint32(pb.DropReason_POLICY_DENIED),
+		DropReasonDesc: pb.DropReason_POLICY_DENIED,
+	}
+
+	_, errs := dfp.OnDecodedFlow(context.TODO(), flow1)
+	assert.Nil(t, errs)
+
+	metricFamilies, err := reg.Gather()
+	require.NoError(t, err)
+	assert.NotNil(t, metricFamilies)
+	assertGatheredDrops(t, metricFamilies)
+
+	// Read in empty config.
+	watcher = metricConfigWatcher{configFilePath: "testdata/valid_empty_metric_cfg.yaml"}
+	cfg, _, _, err = watcher.readConfig()
+	require.Nil(t, err)
+
+	dfp.onConfigReload(context.TODO(), false, 0, *cfg)
+	assert.Nil(t, errs)
+
+	// The existing drop metrics should be removed after the handler is deregistered.
+	metricFamilies, err = reg.Gather()
+	require.NoError(t, err)
+	assert.Nil(t, metricFamilies)
+
+	// Re-register drop handler with same config and collect metrics.
+	watcher = metricConfigWatcher{configFilePath: "testdata/valid_metric_config.yaml"}
```

Comment:
When handler re-registration is attempted with a smaller label set, Prometheus panics with this error:
`"a previously registered descriptor with the same fully-qualified name as Desc{fqName: \"hubble_drop_total\", help: \"Number of drops\", constLabels: {}, variableLabels: {source_pod,source_namespace,reason,protocol}} has different label names or a different help string"` (vendor/github.com/prometheus/client_golang/prometheus/registry.go@315)

Reply:
I don't think that will work because it would mean dropping the values of existing metrics that are unchanged.

Reply:
I want to put the changes to support label updates in a follow-up PR, but use this thread to explore the solutions.
What if put a warning in the docs that changing labels restarts all handlers (equivalent to the current state of restarting cilium with a different static config)?
Do you know what a custom registry implementation would involve to support this?

Reply:
That feels like an unexpected behavior and a potential footman. I'd rather see if we can explore ways to support label updates without those trade-offs.  As far as solving it with a custom registry, that's probably the correct solution, but I'm not sure. A custom collector might work too, but it means changing the metrics handlers more.